runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -1676,38 +1677,94 @@ TR_OpaqueClassBlock *
 TR_J9SharedCacheServerVM::getClassFromSignature(const char * sig, int32_t sigLength, TR_ResolvedMethod * method, bool isVettedForAOT)
    {
    TR_ResolvedRelocatableJ9JITaaSServerMethod* resolvedJITaaSMethod = (TR_ResolvedRelocatableJ9JITaaSServerMethod *)method;
-   return getClassFromSignature(sig, sigLength, (TR_OpaqueMethodBlock *)resolvedJITaaSMethod->getPersistentIdentifier(), isVettedForAOT);
+   TR_OpaqueClassBlock* clazz = nullptr;
+   J9ClassLoader * cl = ((TR_ResolvedJ9Method *)method)->getClassLoader();
+   // If the classloader is the systemClassLoader, which cannot be unloaded,
+   // then look into the global, 'per client', cache
+   void * systemClassLoader = getSystemClassLoader();
+   if (cl == (J9ClassLoader*)systemClassLoader)
+      {
+      PersistentUnorderedMap<std::string, TR_OpaqueClassBlock*> & systemClassByNameMap = _compInfoPT->getClientData()->getSystemClassByNameMap();
+         {
+         OMR::CriticalSection classFromSigCS(_compInfoPT->getClientData()->getSystemClassMapMonitor());
+         auto it = systemClassByNameMap.find(std::string(sig, sigLength));
+         if (it != systemClassByNameMap.end())
+            return it->second;
```

Comment:
I am not 100% sure, but I think that we need to validate whatever comes from the cache. The reason is that this cache is global for AOT and non-AOT queries, so it could have been a non-AOT query that populated the cache and the validation was never done. Maybe ask Irwin.